File: Tests/Test_WaitFor.vb
Option Explicit On
Option Strict On
Option Compare Binary
Option Infer Off

Imports System
Imports System.IO
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace Tests
    Module Tests_WaitFor
        Private Const shell32WindowName As String = "shell32.dll Properties"
        ' https://gist.github.com/Memphizzz/ed69d2500c422019609c
        Private Const shell32WindowClass As String = "#32770"

        Function Test_WaitForWindow1() As Boolean
            If WalkmanLib.ShowProperties(Path.Combine(Environment.SystemDirectory, "shell32.dll")) = False Then
                Return TestString("WaitForWindow1", "ShowProperties returned False", "ShowProperties returned True")
            End If

            Task.Run(Sub()
                         Thread.Sleep(700)
                         SendKeys.SendWait("{ESC}")
                     End Sub)

            Thread.Sleep(400) ' give the window time to show

            Dim result As Boolean = WalkmanLib.WaitForWindow(shell32WindowName, shell32WindowClass, 10)

            Return TestBoolean("WaitForWindow1", result, False)
        End Function

        Function Test_WaitForWindow2() As Boolean
            If WalkmanLib.ShowProperties(Path.Combine(Environment.SystemDirectory, "shell32.dll")) = False Then
                Return TestString("WaitForWindow2", "ShowProperties returned False", "ShowProperties returned True")
            End If

            Task.Run(Sub()
                         Thread.Sleep(700)
                         SendKeys.SendWait("{ESC}")
                     End Sub)

            Thread.Sleep(400)

            Dim waitDone As Boolean = False
            Dim countExited As Boolean = False
            Dim waitTimeout As Integer = 40
            Task.Run(Sub()
                         Console.Write("Waiting for Shell thread to exit. This is expected to take a while, please wait: ")
                         WalkmanLib.ConsoleProgress(0, waitTimeout, waitDone, countExited)
                     End Sub)

            Dim result As Boolean = True
            Try
                result = WalkmanLib.WaitForWindowByThread(shell32WindowName, shell32WindowClass, CType(waitTimeout, UInteger) * 1000UI)
            Finally
                waitDone = True
                Do Until countExited
                    Thread.Sleep(1)
                Loop
                Console.WriteLine()
            End Try

            Return TestBoolean("WaitForWindow2", result, False)
        End Function
    End Module
End Namespace

File: Tests/Test_WaitFor.cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Tests {
    static class Tests_WaitFor {
        private const string shell32WindowName = "shell32.dll Properties";
        // https://gist.github.com/Memphizzz/ed69d2500c422019609c
        private const string shell32WindowClass = "#32770";

        public static bool Test_WaitForWindow1() {
            if (WalkmanLib.ShowProperties(Path.Combine(Environment.SystemDirectory, "shell32.dll")) == false) {
                return GeneralFunctions.TestString("WaitForWindow1", "ShowProperties returned False", "ShowProperties returned True");
            }

            Task.Run(() => {
                Thread.Sleep(700);
                SendKeys.SendWait("{ESC}");
            });

            Thread.Sleep(400); // give the window time to show

            bool result = WalkmanLib.WaitForWindow(shell32WindowName, shell32WindowClass, 10);

            return GeneralFunctions.TestBoolean("WaitForWindow1", result, false);
        }

        public static bool Test_WaitForWindow2() {
            if (WalkmanLib.ShowProperties(Path.Combine(Environment.SystemDirectory, "shell32.dll")) == false) {
                return GeneralFunctions.TestString("WaitForWindow2", "ShowProperties returned False", "ShowProperties returned True");
            }

            Task.Run(() => {
                Thread.Sleep(700);
                SendKeys.SendWait("{ESC}");
            });

            Thread.Sleep(400);

            bool waitDone = false;
            bool countExited = false;
            int waitTimeout = 40;
            if (!Console.IsOutputRedirected) {
                Task.Run(() => {
                    Console.Write("Waiting for Shell thread to exit. This is expected to take a while, please wait: ");
                    WalkmanLib.ConsoleProgress(0, waitTimeout, ref waitDone, ref countExited);
                });
            } else {
                countExited = true;
            }

            bool result = true;
            try {
                result = WalkmanLib.WaitForWindowByThread(shell32WindowName, shell32WindowClass, (uint)waitTimeout * 1000);
            } finally {
                waitDone = true;
                while (!countExited) {
                    Thread.Sleep(1);
                }
                if (!Console.IsOutputRedirected)
                    Console.WriteLine();
            }

            return GeneralFunctions.TestBoolean("WaitForWindow2", result, false);
        }
    }
}
